Draft: Added git OAuth worklfow to RegularUserOcpCheLoginPage.ts

Daybringer opened the pull request:
<!-- Please review the following before submitting a PR:
Che's Contributing Guide: https://github.com/eclipse/che/blob/master/CONTRIBUTING.md
Pull Request Policy: https://github.com/eclipse/che/wiki/Development-Workflow#pull-requests
-->

### What does this PR do?
Implementing git OAuth so the Login page waits on possible 2FA user confirmation.

### Screenshot/screencast of this PR
<!-- Please include a screenshot or a screencast explaining what is doing this PR -->
[success_run.webm](https://github.com/eclipse/che/assets/44302901/f66ad95e-8364-46a2-91a4-5baa4d850ef8)


### What issues does this PR fix or reference?
<!-- Please include any related issue from eclipse che repository (or from another issue tracker).
     Include link to other pull requests like documentation PR from [the docs repo](https://github.com/eclipse/che-docs)
-->
[#22110](https://github.com/eclipse/che/issues/22110)

### How to test this PR?
<!-- Please explain for example :
  - The test platform (openshift, kubernetes, minikube, CodeReady Container, docker-desktop, etc)
  - Installation method: chectl / che-operator
  - steps to reproduce
 -->
- Navigate to `tests/e2e`
- set env variables :
  -  `export TS_SELENIUM_VALUE_OPENSHIFT_OAUTH=true`
  - `export TS_SELENIUM_BASE_URL=https://che-dogfooding.apps.che-dev.x6e0.p1.openshiftapps.com/dashboard/#/`
  - `export TS_SELENIUM_GIT_PROVIDER_USERNAME="your_username"`
  - `export TS_SELENIUM_GIT_PROVIDER_PASSWORD="your_password"`
  - `export TS_SELENIUM_GIT_PROVIDER_OAUTH=true`
  - `export TS_OCP_LOGIN_PAGE_PROVIDER_TITLE=github-app-sre`
- run `npm run test-all-devfiles`


### PR Checklist

[As the author of this Pull Request I made sure that:](https://github.com/eclipse/che/blob/master/CONTRIBUTING.md#pull-request-template-and-its-checklist)

- [x] [The Eclipse Contributor Agreement is valid](https://github.com/eclipse/che/blob/master/CONTRIBUTING.md#the-eclipse-contributor-agreement-is-valid)
- [x] [Code produced is complete](https://github.com/eclipse/che/blob/master/CONTRIBUTING.md#code-produced-is-complete)
- [x] [Code builds without errors](https://github.com/eclipse/che/blob/master/CONTRIBUTING.md#code-builds-without-errors)
- [ ] [Tests are covering the bugfix or new feature ](https://github.com/eclipse/che/blob/master/CONTRIBUTING.md#tests-are-covering-the-bugfix-or-new-feature)
- [x] [The repository devfile is up to date and works](https://github.com/eclipse/che/blob/master/CONTRIBUTING.md#the-repository-devfile-is-up-to-date-and-works)
- [x] [Sections `What issues does this PR fix or reference` and `How to test this PR` completed](https://github.com/eclipse/che/blob/master/CONTRIBUTING.md#sections-what-issues-does-this-pr-fix-or-reference-and-how-to-test-this-pr-completed)
- [ ] [Relevant user documentation updated](https://github.com/eclipse/che/blob/master/CONTRIBUTING.md#relevant-contributing-documentation-updated)
- [ ] [Relevant contributing documentation updated](https://github.com/eclipse/che/blob/master/CONTRIBUTING.md#relevant-contributing-documentation-updated)
- [ ] [CI/CD changes implemented, documented and communicated](https://github.com/eclipse/che/blob/master/CONTRIBUTING.md#cicd-changes-implemented-documented-and-communicated)

### Reviewers

Reviewers, please comment how you tested the PR when approving it.

ScrewTSW commented on tests/e2e/pageobjects/git-providers/OauthPage.ts:
```diff
@@ -17,140 +17,186 @@ import { TimeoutConstants } from '../../constants/TimeoutConstants';
+    await this.driverHelper.waitDisappearance(
+      this.twoFactorInfoBox,
+      TimeoutConstants.TS_SELENIUM_GIT_TWO_FACTOR_TIMEOUT,
+      1000
```
This is probably not necessary since that's the default polling value

ScrewTSW commented on tests/e2e/pageobjects/git-providers/OauthPage.ts:
```diff
@@ -17,140 +17,186 @@ import { TimeoutConstants } from '../../constants/TimeoutConstants';
-    }
+    await this.driverHelper.waitDisappearance(
+      this.twoFactorInfoBox,
+      TimeoutConstants.TS_SELENIUM_GIT_TWO_FACTOR_TIMEOUT,
```
This variable describes number of attempts to use with polling. If you want to use it as a timeout, I'd rather consider calculating it locally in the `waitDisappearanceTwoFactorInfoBox` method as follows:
```
const attempts: number = TimeoutConstants.TS_SELENIUM_GIT_TWO_FACTOR_TIMEOUT / polling;
```
and then use the `attempts` local variable

Daybringer replied:
Thanks for the review, I've taken the polling rate into account and I've rewritten it.

ScrewTSW replied:
ACK, thank you.

ScrewTSW reviewed (commented):
Have you also verified the code with a Che instance where no secondary OAuth provider is set up? (only using Kubernetes/OpenShift OAuth to login)

Daybringer commented:
> Have you also verified the code with a Che instance where no secondary OAuth provider is set up? (only using Kubernetes/OpenShift OAuth to login)

The PR changes apply only when `TS_SELENIUM_GIT_PROVIDER_OAUTH` is set `true` otherwise it continues with OCP login if you meant that.
```
if (TestConstants.TS_SELENIUM_GIT_PROVIDER_OAUTH) {
      await this.oauthPage.login();
    } else {
      await this.ocpLogin.waitOpenShiftLoginWelcomePage();
      await this.ocpLogin.enterUserNameOpenShift(
        TestConstants.TS_SELENIUM_OCP_USERNAME
      );
      await this.ocpLogin.enterPasswordOpenShift(
        TestConstants.TS_SELENIUM_OCP_PASSWORD
      );
      await this.ocpLogin.clickOnLoginButton();
      await this.ocpLogin.waitDisappearanceOpenShiftLoginWelcomePage();
    }
```
If you meant the absence of 2FA on some instances then it will simply fall through when there is none 2FA box waiting to disappear.

nallikaea commented:
Could you please attach screen-cast to see how new login with 2fa should looks like ?
Do you have any logs with successful test run?

nallikaea commented:
@Daybringer Could you, please, fix your code formatting with `lint` script in package.json file?
https://github.com/Daybringer/che/blob/6935fe862bdbd535f10b456a8231760a4fb69d63/tests/e2e/package.json#L7

ScrewTSW reviewed (approved):
LGTM after addressing Maryna's comments

nallikaea commented on tests/e2e/constants/TimeoutConstants.ts:
```diff
@@ -226,6 +226,11 @@ export const TimeoutConstants: any = {
+     * Timeout waiting for disappearance of Two Factor Authentication info box, "360 000" by default
+     */
+    TS_SELENIUM_GIT_TWO_FACTOR_TIMEOUT:
+        Number(process.env.TS_SELENIUM_GIT_TWO_FACTOR_TIMEOUT) || 360_000,
```
Is this timeout should be so long?

nallikaea replied:
You added this timeout into plugin section. Can you move it up, or add some comment?

dmytro-ndp replied:
@Daybringer : timeouts in tests should be as minimal as possible to fail test quickly if there is nothing to wait, and so have test result as soon as possible, and decrease cost of test execution.

Which maximum time of waiting for disappearance of Two Factor Authentication info box did you see when run the test?

Daybringer replied:
@dmytro-ndp Mentioned in https://github.com/eclipse/che/pull/22202/commits/c8885d3dd2977d26e6ae5839ec7eb29e5dc8fdda

nallikaea commented on tests/e2e/pageobjects/git-providers/OauthPage.ts:
```diff
@@ -17,140 +17,186 @@ import { TimeoutConstants } from '../../constants/TimeoutConstants';
 
 @injectable()
 export class OauthPage {
-    private readonly loginForm: By;
```
On our last QE sync call we had an agreement to have all locators as `private static readonly` and By type. It will be very nice if you will fix it in this pageobject class you currently working on ☺️

nallikaea replied:
https://docs.google.com/document/d/1_BlAqY1s8Ou__9iMQR2ZtmUnkuJ1VHDVtSBTV5-_Uv0/edit#heading=h.yr66jvhjgpfu

Daybringer commented:
> @Daybringer Could you, please, fix your code formatting with `lint` script in package.json file? https://github.com/Daybringer/che/blob/6935fe862bdbd535f10b456a8231760a4fb69d63/tests/e2e/package.json#L7

Oops, that's bad on my end, thanks. Didn't run lint after last commit.
Still some inconsistencies in style tho.
For example 2 vs 4 spaces indentation isn't autofixed.
Will elaborate later with attached screencast.

> Could you please attach screen-cast to see how new login with 2fa should looks like ? Do you have any logs with successful test run?

Thanks, added screencast of success run to PR description.

nallikaea commented on tests/e2e/configs/inversify.config.ts:
```diff
@@ -43,37 +44,55 @@ import { OcpMainPage } from '../pageobjects/openshift/OcpMainPage';
 e2eContainer.bind<IDriver>(TYPES.Driver).to(ChromeDriver).inSingletonScope();
-e2eContainer.bind<ITestWorkspaceUtil>(TYPES.WorkspaceUtil).to(TestWorkspaceUtil);
+e2eContainer
+  .bind<ITestWorkspaceUtil>(TYPES.WorkspaceUtil)
```
Could you please explain, for what reason have you add this new lines?

Daybringer replied:
I didn't manually touch the file before and after linting it so this was done by the linter itself.

nallikaea commented on tests/e2e/pageobjects/git-providers/OauthPage.ts:
```diff
@@ -17,140 +17,186 @@ import { TimeoutConstants } from '../../constants/TimeoutConstants';
+          this.submitButton = By.xpath('//*[@value="Sign in"]');
+          this.denyAccessButton = By.xpath('//button[contains(., "Cancel")]');
+          this.twoFactorInfoBox = By.xpath(
+            '//div[contains(@class, "auth-form-header")]'
```
why here added this new line?

nallikaea commented on tests/e2e/pageobjects/login/RegularUserOcpCheLoginPage.ts:
```diff
@@ -20,34 +21,50 @@ import { DriverHelper } from '../../utils/DriverHelper';
+    await this.driverHelper.waitPresence(
+      By.xpath(this.OPEN_SHIFT_LOGIN_LANDING_PAGE_LOCATOR),
+      TimeoutConstants.TS_SELENIUM_LOAD_PAGE_TIMEOUT
+    );
```
![image](https://github.com/eclipse/che/assets/72726250/9e867c4d-94e6-4e78-a6b7-29a3824fdc4a)

I am not sure all this new lines necessary. Can you switch them back?

nallikaea commented on tests/e2e/pageobjects/git-providers/OauthPage.ts:
```diff
@@ -17,140 +17,186 @@ import { TimeoutConstants } from '../../constants/TimeoutConstants';
-        await this.driverHelper.waitVisibility(this.loginForm, TimeoutConstants.TS_SELENIUM_CLICK_ON_VISIBLE_ITEM * 3);
-    }
+    const attempts: number =
+      TimeoutConstants.TS_SELENIUM_GIT_TWO_FACTOR_TIMEOUT /
```
can we move this logic into waitDisappearance() method and pass timeout as time (TimeoutConstants.TS_SELENIUM_GIT_TWO_FACTOR_TIMEOUT)?
Like here:
https://github.com/eclipse/che/blob/666379174db85eecdef7903e47202329ef50a8c7/tests/e2e/utils/DriverHelper.ts#L95

nallikaea commented on tests/e2e/pageobjects/git-providers/OauthPage.ts:
```diff
@@ -17,140 +17,186 @@ import { TimeoutConstants } from '../../constants/TimeoutConstants';
+      await this.clickOnApproveButton();
+      await this.waitDisappearanceOauthPage();
+    } catch (e) {
+      Logger.debug(
```
Why here new lines added?

nallikaea commented on tests/e2e/pageobjects/git-providers/OauthPage.ts:
```diff
@@ -17,140 +17,186 @@ import { TimeoutConstants } from '../../constants/TimeoutConstants';
+    await this.enterPassword(TestConstants.TS_SELENIUM_GIT_PROVIDER_PASSWORD);
+    await this.clickOnLoginButton();
+    await this.waitClosingLoginPage();
+    await this.waitDisappearanceTwoFactorInfoBox();
```
As I can see, 2FA is applied to any scope, but as I remember we don't have it, for example here _https://bitbucket-bitbucket.apps.ds-airgap-v12.crw-qe.com_ or here _https://gitlab.cee.redhat.com_ for factory tests. Did you try to run some test without it? Maybe it is better to add some env variable, like: 
`TestConstants.TS_SELENIUM_IS_2FA_ENABLED === 'true'` ?

Daybringer replied:
I could but the check will fall through when there's no 2FA to wait for so I think another TestConstant isn't required.

Daybringer commented:
> You added this timeout into plugin section. Can you move it up, or add some comment?

---
> @Daybringer : timeouts in tests should be as minimal as possible to fail test quickly if there is nothing to wait, and so have test result as soon as possible, and decrease cost of test execution.
> Which maximum time of waiting for disappearance of Two Factor Authentication info box did you see when run the test?

Fixed with last 2 commits.
The Github 2FA times out exactly after 1 minute so I chose that time as a timeout constant.

Daybringer commented:
>On our last QE sync call we had an agreement to have all locators as private static readonly and By type. It will be very nice if you will fix it in this pageobject class you currently working on
Addressed in last commit. Moved props to static, but they can't have readonly, because exact values are dependent on `GIT_PROVIDER` test constant.

nallikaea commented:
> Addressed in last commit. Moved props to static, but they can't have readonly, because exact values are dependent on `GIT_PROVIDER` test constant.

@Daybringer, actually, they can be readonly if you assign it inside the constructor: https://www.typescriptlang.org/docs/handbook/2/classes.html#readonly
